File: src/cards.ts
export type Suit = 1 | 2 | 3 | 4;

export interface Card {
  rank: number;
  suit: Suit;
  enhancement?: string;
}

const SUITS = ["S", "H", "D", "C"] as const;
const RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"] as const;

export function cardValue(card: Card): number {
  if (card.rank === 14) return 11;
  if (card.rank === 15) return 12;
  if (card.rank > 10) return 10;
  return card.rank;
}

export function rankName(rank: number): string {
  if (rank === 0) return "0";
  if (rank === 14) return "11";
  if (rank === 15) return "12";
  return RANKS[rank - 1];
}

export function suitName(suit: Suit): string {
  return SUITS[suit - 1];
}

export function cardLabel(card: Card): string {
  return rankName(card.rank) + suitName(card.suit);
}

export function cardId(card: Card): string {
  return rankName(card.rank) + "-" + suitName(card.suit);
}

/** Parse "A-S", "10-H", "K-D" etc. into a Card */
export function parseCard(id: string): Card | null {
  const match = id.match(/^([^-]+)-(.+)$/);
  if (!match) return null;
  const rank = rankFromName(match[1]);
  const suitIndex = SUITS.indexOf(match[2] as any) + 1;
  if (rank === null || suitIndex === 0) return null;
  return { rank, suit: suitIndex as Suit };
}

export function rankFromName(name: string): number | null {
  if (name === "A") return 1;
  if (name === "J") return 11;
  if (name === "Q") return 12;
  if (name === "K") return 13;
  if (name === "0") return 0;
  if (name === "11") return 14;
  if (name === "12") return 15;
  const n = parseInt(name, 10);
  return isNaN(n) ? null : n;
}

/** Convenience: make a card from rank number and suit letter */
export function makeCard(rank: number, suit: "S" | "H" | "D" | "C"): Card {
  return { rank, suit: (SUITS.indexOf(suit) + 1) as Suit };
}

// ── Deck helpers ──────────────────────────────────────────────────────────────

function baseRanksForBoard(boardId?: string): number[] {
  const ranks: number[] = [];
  for (let rank = 1; rank <= 13; rank++) ranks.push(rank);
  if (boardId === "Dinosaurs") {
    ranks.push(0);
    ranks.push(14);
    ranks.push(15);
  }
  return ranks;
}

/** Fisher-Yates shuffle using Math.random(). Mutates and returns the deck. */
export function shuffle(deck: Card[]): Card[] {
  for (let i = deck.length - 1; i >= 1; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    const tmp = deck[i];
    deck[i] = deck[j];
    deck[j] = tmp;
  }
  return deck;
}

/**
 * Draw `count` cards from the end of the deck.
 * Returns [drawn, remaining] — does not mutate the original array.
 */
export function draw(deck: Card[], count: number): [Card[], Card[]] {
  const remaining = deck.slice(0, deck.length - count);
  const drawn = deck.slice(deck.length - count);
  return [drawn, remaining];
}

/**
 * Return the default per-card counts for a given board.
 * Each card appears once in a standard deck; the Dinosaurs board adds ranks 0, 14, 15.
 */
export function defaultCounts(boardId?: string): Record<string, number> {
  const counts: Record<string, number> = {};
  const ranks = baseRanksForBoard(boardId);
  for (let suit = 1; suit <= 4; suit++) {
    for (const rank of ranks) {
      const card: Card = { rank, suit: suit as Suit };
      const id = cardId(card);
      counts[id] = (counts[id] ?? 0) + 1;
    }
  }
  return counts;
}

/**
 * Build a full deck of Card objects.
 *
 * Options:
 *   boardId      — selects the rank set (e.g. "Dinosaurs" adds ranks 0/14/15)
 *   deckCounts   — map of cardId → count; overrides the default full deck
 *   enhancements — map of cardId → enhancement string applied to each card
 */
export function buildDeck(options?: {
  boardId?: string;
  deckCounts?: Record<string, number>;
  enhancements?: Record<string, string>;
}): Card[] {
  const boardId = options?.boardId;
  const deckCounts = options?.deckCounts;
  const enhancements = options?.enhancements ?? {};
  const deck: Card[] = [];

  if (deckCounts) {
    for (const [id, count] of Object.entries(deckCounts)) {
      const parsed = parseCard(id);
      if (!parsed) continue;
      for (let n = 0; n < count; n++) {
        const card: Card = { rank: parsed.rank, suit: parsed.suit };
        if (enhancements[id]) card.enhancement = enhancements[id];
        deck.push(card);
      }
    }
  } else {
    const ranks = baseRanksForBoard(boardId);
    for (let suit = 1; suit <= 4; suit++) {
      for (const rank of ranks) {
        const card: Card = { rank, suit: suit as Suit };
        const id = cardId(card);
        if (enhancements[id]) card.enhancement = enhancements[id];
        deck.push(card);
      }
    }
  }

  return deck;
}

File: v4/src/cards.lua
local M = {}

local SUITS = { "S", "H", "D", "C" }
local RANKS = { "A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K" }

local function base_ranks_for_board(board_id)
  local ranks = {}
  for rank = 1, 13 do
    ranks[#ranks + 1] = rank
  end
  if board_id == "Dinosaurs" then
    ranks[#ranks + 1] = 0
    ranks[#ranks + 1] = 14
    ranks[#ranks + 1] = 15
  end
  return ranks
end

local function rank_from_name(rank_name)
  if rank_name == "A" then
    return 1
  end
  if rank_name == "J" then
    return 11
  end
  if rank_name == "Q" then
    return 12
  end
  if rank_name == "K" then
    return 13
  end
  if rank_name == "0" then
    return 0
  end
  if rank_name == "11" then
    return 14
  end
  if rank_name == "12" then
    return 15
  end
  return tonumber(rank_name)
end

local function parse_card_id(card_id)
  local rank_name, suit_name = card_id:match("^([^%-]+)%-(.+)$")
  if not rank_name or not suit_name then
    return nil
  end
  return rank_from_name(rank_name), suit_name
end

function M.build_deck(options)
  local board_id = options and options.board_id or nil
  local deck_counts = options and options.deck_counts or nil
  local enhancements = options and options.enhancements or {}
  local deck = {}
  if deck_counts then
    for id, count in pairs(deck_counts) do
      local rank, suit_name = parse_card_id(id)
      if rank and suit_name then
        local suit_index = nil
        for i = 1, #SUITS do
          if SUITS[i] == suit_name then
            suit_index = i
            break
          end
        end
        if suit_index then
          for _ = 1, count do
            local card = { rank = rank, suit = suit_index }
            if enhancements and enhancements[id] then
              card.enhancement = enhancements[id]
            end
            deck[#deck + 1] = card
          end
        end
      end
    end
  else
    local ranks = base_ranks_for_board(board_id)
    for suit = 1, 4 do
      for i = 1, #ranks do
        local rank = ranks[i]
        local card = { rank = rank, suit = suit }
        local id = M.card_id(card)
        if enhancements and enhancements[id] then
          card.enhancement = enhancements[id]
        end
        deck[#deck + 1] = card
      end
    end
  end
  return deck
end

function M.shuffle(deck)
  for i = #deck, 2, -1 do
    local j = love.math.random(i)
    deck[i], deck[j] = deck[j], deck[i]
  end
  return deck
end

function M.draw(deck, count)
  local drawn = {}
  for _ = 1, count do
    drawn[#drawn + 1] = table.remove(deck)
  end
  return drawn, deck
end

function M.card_label(card)
  return M.rank_name(card.rank) .. SUITS[card.suit]
end

function M.card_value(card)
  if card.rank == 14 then
    return 11
  end
  if card.rank == 15 then
    return 12
  end
  if card.rank > 10 then
    return 10
  end
  return card.rank
end

function M.rank_name(rank)
  if rank == 0 then
    return "0"
  end
  if rank == 14 then
    return "11"
  end
  if rank == 15 then
    return "12"
  end
  return RANKS[rank]
end

function M.card_id(card)
  return M.rank_name(card.rank) .. "-" .. M.suit_name(card.suit)
end

function M.default_counts(board_id)
  local counts = {}
  local ranks = base_ranks_for_board(board_id)
  for suit = 1, 4 do
    for i = 1, #ranks do
      local card = { rank = ranks[i], suit = suit }
      local id = M.card_id(card)
      counts[id] = (counts[id] or 0) + 1
    end
  end
  return counts
end

function M.rank_from_name(name)
  return rank_from_name(name)
end

function M.suit_name(suit)
  return SUITS[suit]
end

return M
